src/lib/parallel_scan/parallel_scan.ml

```diff
@@ -58,24 +59,38 @@ module State = struct
     ; recent_tree_data= []
     ; other_trees_data= []
     ; stateful_work_order= Queue.create ()
-    ; curr_job_seq_no= 0 }
-
-  let transactions
-      { jobs= _
-      ; level_pointer= _
-      ; capacity= _
-      ; acc= _
-      ; current_data_length= _
-      ; base_none_pos= _
-      ; recent_tree_data
-      ; other_trees_data
-      ; stateful_work_order= _
-      ; curr_job_seq_no= _ } =
+    ; curr_job_seq_no= 0
+    ; root_at_depth }
+
+  let transactions {recent_tree_data; other_trees_data; _} =
     List.(rev (concat (recent_tree_data :: other_trees_data)))
 
   let next_leaf_pos p cur_pos =
     if cur_pos = (2 * p) - 2 then p - 1 else cur_pos + 1
 
+  let is_at_root ~root_at_depth pos =
+    let root_starts = Int.pow 2 root_at_depth - 1 in
+    let root_ends = Int.pow 2 (root_at_depth + 1) - 2 in
+    root_starts <= pos && pos <= root_ends
+
+  let slots_per_block state = parallelism state / Int.pow 2 state.root_at_depth
+
+  (*A block is a set of leaves that correspond to a ledger proof. If 
+  root_at_depth = 0 then there would be just one block with all the leaves in 
+  it. No of blocks = 2^root_at_depth. This function returns the starting 
+  position of the block containing pos *)
+  let block_of state pos =
```

Comment:
you could make a pretty easy unit test for this, may as well (especially since you're about to refactor stuff)